arch/armv7a/memcpy.S

```diff
@@ -0,0 +1,361 @@
+#define DST r0
+#define SRC r1
+#define LEN r8
+#define DST_RET r12
+
+.arm
+.syntax unified
+
+.text
+.align 4
+
+/* void *memcpy(void *dst, const void *src, size_t l) */
+.globl memcpy
+.type memcpy, %function
+memcpy:
+	str r8, [sp, #-8]!
+	mov r8, r2
+	cmp LEN, #64
+	mov DST_RET, DST /* preserve return value */
+
+	bhs .LblkCopy
+
+	/* less than 64 bytes - always copy as if block was always unaligned */
+
+.Ltail63Unaligned:
+	/* unaligned copy, 0-63 bytes */
+
+	/* r3 = LEN / 4 */
+	movs r3, LEN, lsr #2
+	beq .Ltail63Un0
+
+.Ltail63Un4:
+	ldr r2, [SRC], #4
+	str r2, [DST], #4
+	subs r3, #1
+	bne .Ltail63Un4
+
+.Ltail63Un0:
+	/* LEN = LEN % 4 */
+	ands LEN, #3
+	beq .Ltail63UnRet
+
+	/* 1 <= LEN <= 3 */
+1:
+	subs LEN, #1
+	ldrb r2, [SRC], #1
+	strb r2, [DST], #1
+	bne 1b
+
+.Ltail63UnRet:
+	ldr r8, [sp], #8
+	mov r0, DST_RET
+	bx lr
+
+
+.LblkCopyUnaligned64:
+	/* src/dst not mutually aligned, more than 64 bytes to copy */
+
+	/* align dst to 64 bytes */
+	add r2, DST, #63
+	bic r2, #63
```

Comment:
Hmm, I don't quite get it. We modify dst (we add some offset), but leave src as is. Wouldn't it cause data to be copied to the wrong place within dst?

Reply:
`Rd` is `r2` here, so it's `r2 = DST + 63`, isn't it?

Reply:
Yeah, misunderstood it completely, sorry